utils/log_parser.py

```diff
@@ -248,36 +251,144 @@ def parse_args() -> Tuple[str, List[str]]:
     else:
         dumpPhases = []
 
-    return options.log_file, dumpPhases
+    if options.db_file:
+        dbFile = options.db_file
+    else:
+        dbFile = "compilation_log_db.sqlite"
+
+    return dbFile, options.log_file, dumpPhases
 
 
 def dump_dag(dagName: str) -> None:
+    """A helper function to dump the DAG."""
+
     dotty = DottyPrinter()
     dotty.dump_graph(dagName)
 
 
-def process(logLines: List[str], dumpPhases: List[str]) -> None:
+def store_transformation_into_DB(
+        transID: int,
+        addedNodes: List[Node],
+        replacedNodes: List[Node],
+        conn: sqlite3.Connection,
+        fullScopeName: str) -> None:
+    """A helper function to store nodes transformations into database.
+
+    Args:
+        transID: int. The ID for this stored transformation.
+        addedNodes: List[Node]. A list of added nodes in this transformation.
+        replacedNodes: List[Node]. A list of replaced nodes in this transformation.
+        conn: sqlite3.Connection. Connection to sqlite3 database
+        fullScopeName: str. The full scope name of this transformation.
+    """
+
+    cursor = conn.cursor()
+    for an in addedNodes:
+        cursor.execute(f"""INSERT INTO Log_Transformation VALUES (
+                        {str(transID)},
+                        'ADD',
+                        '{an.get_name()}',
+                        '{an.get_kind_name()}',
+                        '{fullScopeName}'
+                        )""")
+
+    for rn in replacedNodes:
+        cursor.execute(f"""INSERT INTO Log_Transformation VALUES (
+                        {transID},
+                        'REMOVE',
+                        '{rn.get_name()}',
+                        '{rn.get_kind_name()}',
+                        '{fullScopeName}'
+                        )""")
+
+
+def find_all_replaced_nodes(replacedNode: Node) -> List[Node]:
+    """Find all nodes that will lose user after the given node is removed.
+
+    After one node lost all its uses (e.g. after replaceAllUsesOfWith()), we go through
+    all of its parents to collect all nodes that will consequently lose all their uses.
+
+    Args:
+        replacedNode: Node. The node that just lost all uses.
+    """
+
+    replacedNodeList = []
+    activeDCEList = [replacedNode]
+    while len(activeDCEList):
```

Comment:
It looks like `replacedNodeList` and `activeDCEList` are being used as stacks in this loop. I think the code will be more readable if you use `s.append(item)` and `s.pop()` to push/pop instead of `s = [item] + s` and `s.pop(0)`. The [Python Library Reference](https://docs.python.org/3/tutorial/datastructures.html#using-lists-as-stacks) suggests this pattern as well.